### PR TITLE
Geometry_Engine: Changes to the TaperedProfile constructor and added methods related to TaperedProfiles

### LINKED ISSUE #1978: Geometry_Engine: Add InterpolationOrder to TaperedProfile constructor and determine ShapeProfile
#### Description:

See here for oM issue: https://github.com/BHoM/BHoM/issues/990

### LINKED ISSUE #2014: Structure_Engine: Method to map a TaperedProfile to a List<Bar>
#### Description:

User has a `TaperedProfile` that describes the start and end profile of a `Bar`.

The `Bar` is split as multiple locations to accomodate the `FEMesh` representing the deck and therefore the `TaperedProfile` cannot simply be applied to each `Bar`.

Instead a method will take the `TaperedProfile` and the `List<Bar>` and interpolate the sections for each `Bar` in the `List<Bar>`.

### PR DESCRIPTION
**Pull request by peterjamesnugent**
### NOTE: Depends on 

https://github.com/BHoM/BHoM/pull/992
   
### Issues addressed by this PR

Closes #1978
Closes #2014 

### Test files

https://burohappold.sharepoint.com/:f:/s/BHoM/Eib4yH7k2dNAgoJgo_WZO0cBApcT9JuZC4cOZTK2cjpXMg?e=cT4jcM

### Changelog

- `interpolationOrder` added to constructor for `TaperedProfile` 
- Method to determine the `ShapeType` of the `TaperedProfile` added to constructor, this will be `FreeForm` if the `ShapeType` is not the same
- `InterpolateProfile` method added to the Geometry_Engine
-  `MapTaperedProfile` added to Structure_Engine to allow mapping of a `TaperedProfile` to a `List<Bar>` interpolating the intermediate `TaperedProfile` for each `Bar`
- Several methods added such as `MapDomain`, `MapPositionDomain` added to aid `MapTaperedProfile`

### CODE REVIEW

**Comment by peterjamesnugent** [comment]
@IsakNaslundBh as discussed the `positions` returned from `MapTaperedProfile()` are relative to the `TaperedProfile` they have been derived from. This will mean that any further interpolation will be correct. 

I've also added a `MapPositionDomain` function that allows a domain to be mapped to `[0,1]` for when it's being pushed in to an FEA software.

File: Geometry_Engine/Compute/InterpolateProfile.cs [hunk 1]
```diff
@@ -0,0 +1,230 @@
...
+ */
+
+using System;
+using System.Collections.Generic;
+using BH.oM.Geometry;
+using BH.oM.Geometry.ShapeProfiles;
+using System.ComponentModel;
+using BH.oM.Reflection.Attributes;
+
+namespace BH.Engine.Geometry
```
Comment (IsakNaslundBh) [inline Geometry_Engine/Compute/InterpolateProfile.cs:30]: Would move this whole class to spatial_Engine and namespace. We are planning to migrate all profiles over there early this milestone, preferably next sprint, and would like to have as little versioning as possible! :)

File: Geometry_Engine/Compute/InterpolateProfile.cs [hunk 2]
```diff
@@ -0,0 +1,230 @@
...
+        /**** Public Methods                            ****/
+        /***************************************************/
+
+        [Description("Compute an IProfile by using parabolic interpolation at a given parameter between two IProfile objects of the same type.")]
+        [Input("startProfile", "The IProfile at the start.")]
+        [Input("endProfile", "The IProfile at the end.")]
+        [Input("parameter", "A number between 0 and 1 that describes the distance along the line.")]
+        [Input("interpolationOrder", "The value of the polynomimal function used the describe the transition between the startProfile and endProfile.")]
+        [Output("interpolatedProfile", "The profile evaluated at the parameter given using interpolation between the startProfile and endProfile using a function with the given interpolation order.")]
+        public static IProfile InterpolateProfile(IProfile startProfile, IProfile endProfile, double parameter, int interpolationOrder)
```
Comment (IsakNaslundBh) [inline Geometry_Engine/Compute/InterpolateProfile.cs:44]: Should be an interface method. Would also move this to the bottom of the file under a 

```
        /***************************************************/
        /**** Public Methods - Interfaces               ****/
        /***************************************************/
```

```suggestion
        public static IProfile IInterpolateProfile(IProfile startProfile, IProfile endProfile, double parameter, int interpolationOrder)
```


Should also add a private fallback method accepting two `IProfile`s (same signature that this currently have) raising an error and returning null.

File: Geometry_Engine/Compute/InterpolateProfile.cs [hunk 3]
```diff
@@ -0,0 +1,230 @@
...
+        {
+            return InterpolateProfile(startProfile as dynamic, endProfile as dynamic, parameter, interpolationOrder);
+        }
+
+        /***************************************************/
+
+        private static AngleProfile InterpolateProfile(AngleProfile startProfile, AngleProfile endProfile, double parameter, int interpolationOrder)
```
Comment (IsakNaslundBh) [inline Geometry_Engine/Compute/InterpolateProfile.cs:51]: Should be public to follow general BHoM standard. Same for all the other methods.

Understand why you have done it like this from the side of the UI, but think we should follow what we are doing elsewhere atm.

File: Geometry_Engine/Modify/MApPositionDomain.cs [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+ */
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.ComponentModel;
+using BH.oM.Reflection.Attributes;
+using BH.oM.Geometry.ShapeProfiles;
+
+namespace BH.Engine.Geometry
```
Comment (IsakNaslundBh) [inline Geometry_Engine/Modify/MApPositionDomain.cs:52]: Same as above, lets put this in Spatial_Engine straight away.

File: Structure_Engine/Compute/MapTaperedProfile.cs [hunk 5]
```diff
@@ -0,0 +1,267 @@
...
+        /***************************************************/
+        /**** Public Methods                            ****/
+        /***************************************************/
+
+        [Description("Maps a TaperedProfile to a series of sequential Bars. For example, the start and end profiles are known for the series of bars, this method " +
+            "will interpolate between bars to create a TaperedProfile for each Bar.")]
+        [Input("section", "The section containing the TaperedProfile to be mapped to the series of Bars")]
+        [Input("bars", "The Bars for the TaperedProfile to be mapped to.")]
+        [Output("bars", "The Bars with interpolated SectionProperties based on the TaperedProfile provided.")]
+        public static List<Bar> IMapTaperedSection(List<Bar> bars, ISectionProperty section)
```
Comment (IsakNaslundBh) [inline Structure_Engine/Compute/MapTaperedProfile.cs:48]: Would put this at the bottom of the file under a

```
        /***************************************************/
        /**** Public Methods - Interfaces               ****/
        /***************************************************/
```

File: Structure_Engine/Compute/MapTaperedProfile.cs [hunk 6]
```diff
@@ -0,0 +1,267 @@
...
+                    sections[i].Name = section.Name + "-s" + i;
+                    newBars[i].SectionProperty = sections[i];
+                }
+            }
+
+            return newBars;
+        }
+
+        /***************************************************/
+        public static List<Bar> MapTaperedSection(List<Bar> bars, ISectionProperty section)
```
Comment (IsakNaslundBh) [inline Structure_Engine/Compute/MapTaperedProfile.cs:82]: Would make this private as it is kind of a fallback.

File: Structure_Engine/Compute/MapTaperedProfile.cs [hunk 7]
```diff
@@ -0,0 +1,267 @@
...
+                        int newInterpolationOrder;
+                        try
+                        {
+                            newInterpolationOrder = newInterpolationOrders[i];
+                        }
+                        catch (ArgumentOutOfRangeException)
+                        {
+                            //If it is the final index
+                            newInterpolationOrder = newInterpolationOrders[i - 1];
+                        }
```
Comment (IsakNaslundBh) [inline Structure_Engine/Compute/MapTaperedProfile.cs:216]: We should not use try-catch as a filter mechanism. Exceptions should only be used for real exceptions. They are significantly slower than doing the check needed ourselfs. Reading the comment of the code, I think something along the lines the following could work:


```suggestion
                        int interpolationIndex = Math.Min(i, newInterpolationOrders.Count-1);
                        int newInterpolationOrder = newInterpolationOrders[interpolationIndex];

```

Good to add some comment as well as to how the `i` relates to the list of interpolation orders. Assume you have one less interpolation order than number of profiles?

**Review by IsakNaslundBh — CHANGES_REQUESTED** [review]
Some first review of just the code here on github. Some pre-migration of stuff out of geometry_Engine (new stuff added) and some general smaller compliance things.

File: Geometry_Engine/Create/ShapeProfile.cs [hunk 8]
```diff
@@ -554,44 +554,66 @@ public static TubeProfile TubeProfile(double diameter, double thickness)
 
         /***************************************************/
 
-        public static TaperedProfile TaperedProfile(List<decimal> positions, List<IProfile> profiles)
+        [PreviousVersion("3.3", "BH.Engine.Geometry.Create.TaperedProfile(System.Collections.Generic.List<System.Decimal>, System.Collections.Generic.List<BH.oM.Geometry.ShapeProfiles.IProfile>)")]
```
Comment (IsakNaslundBh) [inline Geometry_Engine/Create/ShapeProfile.cs:557]:
```suggestion
        [PreviousVersion("4.0", "BH.Engine.Geometry.Create.TaperedProfile(System.Collections.Generic.List<System.Decimal>, System.Collections.Generic.List<BH.oM.Geometry.ShapeProfiles.IProfile>)")]
```

File: Geometry_Engine/Create/ShapeProfile.cs [hunk 9]
```diff
@@ -554,44 +554,66 @@ public static TubeProfile TubeProfile(double diameter, double thickness)
...
+            ShapeType shape = GetShapeType(profiles);
+            TaperedProfile taperedProfile = new TaperedProfile(profileDict, interpolationOrder, shape);
+
+            return taperedProfile;
         }
 
         /***************************************************/
 
-        public static TaperedProfile TaperedProfile(IProfile startProfile, IProfile endProfile)
+        [PreviousVersion("3.3", "BH.Engine.Geometry.Create.TaperedProfile(BH.oM.Geometry.ShapeProfiles.IProfile, BH.oM.Geometry.ShapeProfiles.IProfile)")]
```
Comment (IsakNaslundBh) [inline Geometry_Engine/Create/ShapeProfile.cs:612]:
```suggestion
        [PreviousVersion("4.0", "BH.Engine.Geometry.Create.TaperedProfile(BH.oM.Geometry.ShapeProfiles.IProfile, BH.oM.Geometry.ShapeProfiles.IProfile)")]
```

**Review by IsakNaslundBh — CHANGES_REQUESTED** [review]
Wrong version in previous method attribute. Should be the current version we are working ok :)

Not sure if there are more that I missed

**Review by IsakNaslundBh — APPROVED** [review]
Seem to be working fine now! Great to get the interpolate methods added!

**Pull request merged**